File: test/tokens.test.ts
import { expect, test } from "bun:test";
import { Tokens } from "../src/cp/tokens.ts";

test("signs and verifies by kind", () => {
  const t = new Tokens("secret");
  const tok = t.sign({ k: "attach", sid: "s_abc" }, 1000);
  expect(t.verify(tok, "attach")?.sid).toBe("s_abc");
  expect(t.verify(tok, "preview")).toBeNull();
});

test("rejects tampering, wrong secret, expiry", () => {
  const t = new Tokens("secret");
  const tok = t.sign({ k: "preview", sid: "s_abc", port: 3000 }, 1000);
  expect(new Tokens("other").verify(tok, "preview")).toBeNull();
  const [body, sig] = tok.split(".");
  expect(t.verify(`${body}x.${sig}`, "preview")).toBeNull();
  expect(t.verify(`${body}.${sig!.slice(0, -1)}A`, "preview")).toBeNull();
  expect(t.verify(t.sign({ k: "preview", sid: "s_abc" }, -1), "preview")).toBeNull();
  expect(t.verify(null, "preview")).toBeNull();
  expect(t.verify("garbage", "preview")).toBeNull();
});

File: internal/cp/tokens_test.go
package cp

import (
	"strings"
	"testing"
	"time"
)

func TestTokensSignAndVerifyByKind(t *testing.T) {
	tk := NewTokens("secret")
	token := tk.Sign(TokenPayload{Kind: TokenAttach, SID: "s_abc"}, time.Second)

	p, ok := tk.Verify(token, TokenAttach)
	if !ok || p.SID != "s_abc" {
		t.Fatalf("verify = %+v, %v", p, ok)
	}
	// The kind is part of what the signature means: a preview cookie is not an attach token.
	if _, ok := tk.Verify(token, TokenPreview); ok {
		t.Error("a token verified as the wrong kind")
	}
}

func TestTokensReject(t *testing.T) {
	tk := NewTokens("secret")
	token := tk.Sign(TokenPayload{Kind: TokenPreview, SID: "s_abc", Port: 3000}, time.Second)
	body, sig, _ := strings.Cut(token, ".")

	cases := map[string]struct {
		tokens *Tokens
		token  string
	}{
		"another secret":  {NewTokens("other"), token},
		"tampered body":   {tk, body + "x." + sig},
		"tampered sig":    {tk, body + "." + flipFirst(sig)},
		"no signature":    {tk, body},
		"empty":           {tk, ""},
		"garbage":         {tk, "garbage"},
		"already expired": {tk, tk.Sign(TokenPayload{Kind: TokenPreview, SID: "s_abc"}, -time.Second)},
	}
	for name, c := range cases {
		t.Run(name, func(t *testing.T) {
			if _, ok := c.tokens.Verify(c.token, TokenPreview); ok {
				t.Error("accepted")
			}
		})
	}
}

// flipFirst changes the leading character to a different one, so a signature is always
// altered — picking a fixed letter would leave it untouched one time in sixty-four.
//
// The first and not the last: an HMAC-SHA256 is 32 bytes, which base64url encodes as 43
// characters carrying 258 bits, so the final character's low two bits are padding that the
// decoder discards. Changing "A" to "B" there decodes to the very same 32 bytes, and the
// case verified rather than failed about one run in sixteen.
func flipFirst(s string) string {
	if s[0] == 'A' {
		return "B" + s[1:]
	}
	return "A" + s[1:]
}

func TestTokensCarryThePort(t *testing.T) {
	tk := NewTokens("k")
	token := tk.Sign(TokenPayload{Kind: TokenPreviewCookie, SID: "s_1", Port: 8888}, time.Minute)
	p, ok := tk.Verify(token, TokenPreviewCookie)
	if !ok || p.Port != 8888 {
		t.Errorf("verify = %+v, %v", p, ok)
	}
	// The signature says the token is ours, not that it is for the sandbox in front of us:
	// the caller compares sid and port itself, which is why they must survive the round trip.
	if p.SID != "s_1" {
		t.Errorf("sid = %q", p.SID)
	}
}
